File: app/src/main/java/org/zeroxlab/momodict/Controller.java
package org.zeroxlab.momodict;

import android.content.Context;
import android.support.annotation.NonNull;

import org.zeroxlab.momodict.db.realm.RealmStore;
import org.zeroxlab.momodict.model.Book;
import org.zeroxlab.momodict.model.Card;
import org.zeroxlab.momodict.model.Entry;
import org.zeroxlab.momodict.model.Record;
import org.zeroxlab.momodict.model.Store;

import java.util.Collections;
import java.util.List;

import rx.Observable;

public class Controller {

    private Context mCtx;
    private Store mStore;

    public Controller(@NonNull Context ctx) {
        mCtx = ctx;
        mStore = new RealmStore(mCtx);
    }

    public Observable<Book> getBooks() {
        return Observable.from(mStore.getBooks());
    }

    public Observable<Entry> queryEntries(String keyWord) {
        List<Entry> list = mStore.queryEntries(keyWord);
        Collections.sort(list, (left, right) -> {
            return left.wordStr.indexOf(keyWord) - right.wordStr.indexOf(keyWord);
        });
        return Observable.from(list);
    }

    public Observable<Entry> getEntries(String keyWord) {
        List<Entry> list = mStore.getEntries(keyWord);
        Collections.sort(list, (left, right) -> {
            return left.wordStr.indexOf(keyWord) - right.wordStr.indexOf(keyWord);
        });
        return Observable.from(list);
    }

    public Observable<Record> getRecords() {
        List<Record> records = mStore.getRecords();
        Collections.sort(records, (left, right) -> {
            // sorting by time. Move latest one to head
            return left.time.before(right.time) ? 1 : -1;
        });
        return Observable.from(records);
    }

    public void clearRecords() {
        getRecords().subscribe(
                (record -> mStore.removeRecords(record.wordStr)));
    }

    public boolean setRecord(@NonNull Record record) {
        return mStore.upsertRecord(record);
    }

    public boolean removeRecord(@NonNull String keyWord) {
        return mStore.removeRecords(keyWord);
    }

    public Observable<Card> getCards() {
        List<Card> cards = mStore.getCards();
        Collections.sort(cards, (left, right) -> {
            // sorting by time. Move latest one to head
            return left.time.before(right.time) ? 1 : -1;
        });
        return Observable.from(cards);
    }

    public boolean setCard(@NonNull Card card) {
        return mStore.upsertCard(card);
    }

    public boolean removeCards(@NonNull String keyWord) {
        return mStore.removeCards(keyWord);
    }
}

File: app/src/main/java/org/zeroxlab/momodict/Controller.kt
package org.zeroxlab.momodict

import android.content.Context
import androidx.room.Room.databaseBuilder
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.zeroxlab.momodict.db.room.RoomStore
import org.zeroxlab.momodict.model.Book
import org.zeroxlab.momodict.model.Card
import org.zeroxlab.momodict.model.Entry
import org.zeroxlab.momodict.model.Record
import org.zeroxlab.momodict.model.Store

// FIXME: should avoid main thread
// TODO: How to handle exception in each function call?
class Controller @JvmOverloads constructor(
    private val mCtx: Context,
    private val mStore: Store = databaseBuilder(
        mCtx.applicationContext,
        RoomStore::class.java,
        RoomStore.DB_NAME
    )
        .allowMainThreadQueries()
        .build()
) {

    // sorting by time. Move latest one to head
    private val recordTimeComparator: Comparator<Record> = Comparator { left, right ->
        if (left.time!!.before(right.time)) 1 else -1
    }

    // sorting by time. Move latest one to head
    private val cardTimeComparator = Comparator<Card> { left, right ->
        if (left.time!!.before(right.time)) 1 else -1
    }

    fun getBooks(scope: CoroutineScope, cb: (List<Book>) -> Unit) {
        scope.launch(Dispatchers.IO) {
            val books = mStore.getBooks()
            withContext(scope.coroutineContext) {
                cb(books)
            }
        }
    }

    fun removeBook(bookName: String): Boolean {
        return mStore.removeBook(bookName)
    }

    suspend fun queryEntries(
        keyWord: String
    ): List<Entry> = withContext(Dispatchers.IO) {
        // to make sure exact matched words are returned
        val exact = syncGetEntries(keyWord)
        val list = mStore.queryEntries(keyWord)
        val comparator = Comparator<Entry> { left, right ->
            left.wordStr!!.indexOf(keyWord) - right.wordStr!!.indexOf(keyWord)
        }

        list.sortWith(comparator)
        exact.forEach { list.add(0, it) }
        val distinct = list.distinctBy { item -> item.wordStr }
        distinct
    }

    suspend fun getEntries(keyWord: String): List<Entry> = withContext(Dispatchers.IO) {
        // to make sure exact matched words are returned
        syncGetEntries(keyWord)
    }

    suspend fun getRecords(): (List<Record>) = withContext(Dispatchers.IO) {
        mStore.getRecords().apply { sortWith(recordTimeComparator) }
    }

    suspend fun clearRecords() = withContext(Dispatchers.IO) {
        val records = getRecords()
        records.forEach { r ->
            mStore.removeRecords(r.wordStr)
        }
    }

    suspend fun setRecord(record: Record): Boolean = withContext(Dispatchers.IO) {
        mStore.upsertRecord(record)
    }

    suspend fun removeRecord(keyWord: String): Boolean = withContext(Dispatchers.IO) {
        mStore.removeRecords(keyWord)
    }

    suspend fun getCards(): List<Card> = withContext(Dispatchers.IO) {
        val cards = mStore.getCards()
        cards.sortWith(cardTimeComparator)
        cards
    }

    suspend fun setCard(card: Card): Boolean = withContext(Dispatchers.IO) {
        mStore.upsertCard(card)
    }

    suspend fun removeCards(keyWord: String): Boolean = withContext(Dispatchers.IO) {
        mStore.removeCards(keyWord)
    }

    private fun syncGetEntries(keyWord: String): List<Entry> {
        val list = mStore.getEntries(keyWord)
        val comparator = Comparator<Entry> { left, right ->
            left.wordStr!!.indexOf(keyWord) - right.wordStr!!.indexOf(keyWord)
        }
        list.sortWith(comparator)
        return list
    }
}
